demo/lib/distro.bash

```diff
@@ -25,40 +27,61 @@ distro-remove-pkg()         { distro-resolve "$@"; }
 distro-setup-proxies()      { distro-resolve "$@"; }
 distro-install-golang()     { distro-resolve "$@"; }
 distro-install-containerd() { distro-resolve "$@"; }
+distro-config-containerd()  { distro-resolve "$@"; }
 distro-install-crio()       { distro-resolve "$@"; }
 distro-install-k8s()        { distro-resolve "$@"; }
+distro-k8s-cni()            { distro-resolve "$@"; }
 distro-set-kernel-cmdline() { distro-resolve "$@"; }
 distro-bootstrap-commands() { distro-resolve "$@"; }
 
-# default no-op fallbacks for optional API functions
-default-bootstrap-commands() { :; }
-
 ###########################################################################
 
 # distro-specific function resolution
 distro-resolve() {
-    # We dig out the distro-* API function name from stack of callers,
-    # then try resolving it to an implementation. The resolution
-    # process goes through a list of potential implementation names in
-    # decreasing order of distro/version-specificity, then tries a few
-    # fallbacks based on known distro relations and properties.
-    local apifn="${FUNCNAME[1]}" fallbacks fn
+    local apifn="${FUNCNAME[1]}" fn prefn postfn
+    # shellcheck disable=SC2086
+    {
+        fn="$(distro-resolve-fn $apifn)"
+        prefn="$(distro-resolve-fn $apifn-pre)"
+        postfn="$(distro-resolve-fn $apifn-post)"
+    }
+
```

Comment:
I agree that sometimes it might/will feel counter-intuitive which hooks are run. Was considering but (out of laziness) didn't as some of these functions are evaluated with $()-output substitution so we need to be careful about what is getting logged and where. But better add it now than find ourselves be chasing ghosts for hours later...

Reply:
Added deubg-log()/logging to $OUTPUT_DIR/debug-log.